docs/modules/demos/pages/jupyterhub-keycloak.adoc

```diff
@@ -0,0 +1,189 @@
+
+For this demo a small sample of {gas-sensor}[gas sensor measurements*] is provided.
+Install this demo on an existing Kubernetes cluster:
+
+[source,console]
+----
+$ stackablectl demo install jupyterhub-keycloak
+----
+
+WARNING: When running a distributed Spark cluster from within a JupyterHub notebook, the notebook acts as the driver and requests executors Pods from k8s.
+These Pods in turn can mount *all* volumes and Secrets in that namespace.
+To prevent this from breaking user separation, it is planned to use an OPA gatekeeper to define OPA rules that restrict what the created executor Pods can mount. This is not yet implemented in this demo.
+
+[#system-requirements]
+== System requirements
+
+To run this demo, your system needs at least:
+
+* 8 {k8s-cpu}[cpu units] (core/hyperthread)
+* 32GiB memory
+
+You may need more resources depending on how many concurrent users are logged in, and which notebook profiles they are using.
+
+== Aim / Context
+
+This demo shows how to authenticate JupyerHub users against a Keycloak backend using JupyterHub's OAuthenticator.
+The same users as in the xref:end-to-end-security.adoc[End-to-end-security] demo are configured in Keycloak and these will be used as examples.
+The notebook offers a simple template for using Spark to interact with S3 as a storage backend.
+
+== Overview
+
+This demo will:
+
+* Install the required Stackable Data Platform operators
+* Spin up the following data products:
+** *JupyterHub*: A multi-user server for Jupyter notebooks
+** *Keycloak*: An identity and access management product
+** *S3*: A Minio instance for data storage
+* Download a sample of the gas sensor dataset into S3
+* Install the Jupyter notebook
+* Demonstrate some basic data operations against S3
+* Illustrate multi-user usage
+
+== JupyterHub
+
+Have a look at the available Pods before logging in:
+
+[source,console]
+----
+$ kubectl get pods
+NAME                         READY   STATUS      RESTARTS   AGE
+hub-84f49ccbd7-29h7j         1/1     Running     0          56m
+keycloak-544d757f57-f55kr    2/2     Running     0          57m
+load-gas-data-m6z5p          0/1     Completed   0          54m
+minio-5486d7584f-x2jn8       1/1     Running     0          57m
+proxy-648bf7f45b-62vqg       1/1     Running     0          56m
+
+----
+
+The `proxy` Pod has an associated `proxy-public` service with a statically-defined port (31095), exposed with type NodePort.
+In order to reach the JupyterHub web interface, navigate to this service.
+The node port IP can be found in the ConfigMap `keycloak-address` (written by the Keycloak Deployment as it starts up).
+In the example below that would then be 172.19.0.5:31095:
+
+[source,yaml]
+----
+apiVersion: v1
+data:
+  keycloakAddress: 172.19.0.5:31093 # Keycloak itself
+  keycloakNodeIp: 172.19.0.5 # can be used to access the proxy-public service
```

Comment:
What if its on another node?

Reply:
I've just checked - you can reach the jupyterhub proxy-public from any node.

Reply:
This may be a KinD thing, though. Not sure.

Reply:
I'll add a comment.

Reply:
See 7be5288